pkgs/development/libraries/flutter-engine/package.nix

```diff
@@ -0,0 +1,177 @@
+## RFC: should we use "hostPackages" to get the packages we need for "nativeBuildInputs"?
+{ lib, stdenv, stdenvNoCC, hostPlatform, callPackage, fetchFromGitHub, fetchurl, writeText,
+  ninja, gnumake, patchelf, python3, clang-tools, pkg-config, openssh, git, gclient-wrapped }:
+{ sha256, version, runtimeMode }:
+with lib;
+let
+  flutter-deps = filterAttrs (name: pkg: isAttrs pkg && hasAttr "outPath" pkg) (callPackage ./deps.nix {});
+
+  toolchainArch = if hostPlatform.isx86_64 then "amd64" else if hostPlatform.isAarch64 then "aarch64" else throws "Unsupported platform";
+  interpreter = "ld-linux-${hostPlatform.parsed.cpu.arch}.so.2";
+
+  src = stdenvNoCC.mkDerivation {
+    pname = "flutter-engine-src";
+    inherit version;
+
+    passthru = {
+      inherit flutter-deps;
+    };
+
+    nativeBuildInputs = [ patchelf python3 ];
+
+    src = fetchFromGitHub {
+      owner = "flutter";
+      repo = "engine";
+      rev = version;
+      inherit sha256;
+    };
+
+    buildroot = fetchFromGitHub {
+      owner = "flutter";
+      repo = "buildroot";
+      rev = "6af51ff4b86270cc61517bff3fff5c3bb11492e1";
+      sha256 = "sha256-03jr3H1RvzGfa0rBPZ1rtNpmieKzDjDgBsrZGaj7vuI=";
+    };
+
+    toolchain = fetchurl {
+      url = "https://commondatastorage.googleapis.com/chrome-linux-sysroot/toolchain/79a7783607a69b6f439add567eb6fcb48877085c/debian_sid_${toolchainArch}_sysroot.tar.xz";
+      sha256 = "sha256-+jooJ4YcWHUfD5AT+Yhe0nYQNBl7IvJl2dRc+d3Yl+g=";
+    };
+
+    dontUnpack = true;
+    dontConfigure = true;
+    dontBuild = true;
+
+    ## FIXME: use "flutter-deps" and the arguments which the OG script takes to generate the git revision
+    gitrev = writeText "git-revision.py" ''
+      #!${python3}/bin/python3
+
+      print("${fakeHash}")
+    '';
+
+    dartPackageConfig = writeText "dart-package-config.json" ''
+      {
+        "configVersion": 2,
+        "generator": "tools/generate_package_config.dart",
+        "packages": []
+      }
+    '';
+
+    binaryFixes = [
+      "src/third_party/dart/tools/sdks/dart-sdk/bin/dart"
+      "src/flutter/third_party/gn/gn"
+    ];
+
+    shellFixes = [
+      "src/third_party/dart/tools/sdks/dart-sdk/bin/dart2js"
+      "src/flutter/tools/gn"
+    ];
+
+    ## FIXME: instead of patching a bunch of binaries, maybe we copy from nixpkgs?
+    installPhase = ''
+      mkdir -p $out
+      cp -r -P --no-preserve=mode,ownership $buildroot $out/src
```

Comment:
remove all the `--no-preserve=mode,ownership` and use `chmod -R +w` instead, to preserve executable bits.
remove binaryFixes
remove shellFixes. fixupPhase should run patchShebangs on $out

Reply:
We have to patch those sooner because the python scripts require running those executables and the patch hooks don't catch it for some reason.